classes/user/Collector.php

```diff
@@ -457,42 +457,61 @@ protected function buildUserGroupFilter(Builder $query): self
+                    ->whereIn('ugs.stage_id', $this->workflowStageIds)
+            )
+            ->when($this->roleIds !== null, fn ($subQuery) => $subQuery->whereIn('ug.role_id', $this->roleIds))
+            ->when($this->contextIds !== null, fn ($subQuery) => $subQuery->whereIn('ug.context_id', $this->contextIds))
+            ->when(
+                $this->userUserGroupStatus === UserUserGroupStatus::STATUS_ACTIVE,
+                fn (Builder $subQuery) =>
+                $subQuery->where(
+                    fn (Builder $subQuery) =>
+                    $subQuery->where('uug.date_start', '<=', $currentDateTime)
+                        ->orWhereNull('uug.date_start')
                 )
-                ->when($this->excludeRoles !== null, fn ($query) => $query->whereNotIn('ug.role_id', $this->excludeRoles))
-                ->when($this->roleIds !== null, fn ($query) => $query->whereIn('ug.role_id', $this->roleIds))
-                ->when($this->contextIds !== null, fn ($query) => $query->whereIn('ug.context_id', $this->contextIds))
-                ->when(
-                    $this->userUserGroupStatus === UserUserGroupStatus::STATUS_ACTIVE,
-                    fn (Builder $query) =>
-                    $query->where(
-                        fn (Builder $query) =>
-                        $query->where('uug.date_start', '<=', $currentDateTime)
-                            ->orWhereNull('uug.date_start')
+                    ->where(
+                        fn (Builder $subQuery) =>
+                        $subQuery->where('uug.date_end', '>', $currentDateTime)
+                            ->orWhereNull('uug.date_end')
                     )
-                        ->where(
-                            fn (Builder $query) =>
-                            $query->where('uug.date_end', '>', $currentDateTime)
-                                ->orWhereNull('uug.date_end')
-                        )
-                )
-                ->when(
-                    $this->userUserGroupStatus === UserUserGroupStatus::STATUS_ENDED,
-                    fn (Builder $query) =>
-                    $query->whereNotNull('uug.date_end')
-                        ->where('uug.date_end', '<=', $currentDateTime)
+            )
+            ->when(
+                $this->userUserGroupStatus === UserUserGroupStatus::STATUS_ENDED,
+                fn (Builder $subQuery) =>
+                $subQuery->whereNotNull('uug.date_end')
+                    ->where('uug.date_end', '<=', $currentDateTime)
+            );
+
+        $joinSub = clone $subQuery;
+        $query
+            ->whereExists(
+                $subQuery->when(
+                    $this->excludeRoles != null,
+                    // This aggregates a column role_count, which holds an information whether user holds specified role
+                    fn (Builder $subQuery) => $subQuery->leftJoinSub(
```

Comment:
What do you think about using this here instead:
```
fn (Builder $subQuery) => $subQuery->whereNotIn(
  'u.user_id',
  $joinSub
    ->select('uug.user_id')
    ->whereIn('ug.role_id', $this->excludeRoles)
)
```

I wanted to double check and test that tomorrow, but wanted to write it already to you too, in case... :-)

Reply:
Hi @Vitaliy-1, I believe the following would be the best query to use:
```
                    fn (Builder $subQuery) => $subQuery->leftJoinSub(
                        $joinSub
                            ->select('uug.user_id')
                            ->whereIn('ug.role_id', $this->excludeRoles),
                        'agr',
                        'uug.user_id',
                        'agr.user_id'
                    )->whereNull('agr.user_id')

```

What do you think?

Reply:
Agree!